src/main.nr

```diff
@@ -1,24 +1,155 @@
 mod types;
- 
+
 contract CoinToss {
     use dep::aztec::context::Context;
+    use crate::types::bet_note::{BetNote, BetNoteMethods, BET_NOTE_LEN};
+    use crate::types::bet_amount_note::{BetAmountNote, BetAmountNoteMethods, BET_AMOUNT_NOTE_LEN};
+    use crate::types::address_note::{AddressNote, AddressNoteMethods, ADDRESS_NOTE_LEN};
+
+    use dep::aztec::note::{
+        note_getter_options::NoteGetterOptions,
+        note_header::NoteHeader,
+        note_interface::NoteInterface,
+        note_viewer_options::NoteViewerOptions,
+        utils as note_utils
+    };
+
+    use dep::std::option::Option;
+
+    use dep::aztec::log::emit_encrypted_log;
+    use dep::aztec::oracle::{
+        get_public_key::get_public_key,
+        rand::rand
+    };
+    use dep::aztec::state_vars::{
+        public_state::PublicState,
+        map::Map,
+        set::Set,
+        singleton::Singleton,
+        immutable_singleton::ImmutableSingleton
+    };
+    use dep::aztec::selector::compute_selector;
+
+    use dep::aztec::types::type_serialization::field_serialization::{ FieldSerializationMethods, FIELD_SERIALIZED_LEN};
+    use dep::aztec::types::type_serialization::aztec_address_serialization::{AztecAddressSerializationMethods, AZTEC_ADDRESS_SERIALIZED_LEN};
 
+    use dep::aztec::types::address::AztecAddress;
+
+
+    global BET_AMOUNT_SLOT = 1;
+    global DIVINITY_ADDRESS_SLOT = 2;
+    global PRIVATE_ORACLE_ADDRESS_SLOT = 3;
+    global HOUSE_ADDRESS_SLOT = 4;
+    global BETS_SLOT = 5;
+
+    // TODO: add slow update tree to make the switch from private to public
     struct Storage {
+        bet_amount: ImmutableSingleton<BetAmountNote, BET_AMOUNT_NOTE_LEN>,
+        divinity: ImmutableSingleton<AddressNote, ADDRESS_NOTE_LEN>,
+        private_oracle: ImmutableSingleton<AddressNote, ADDRESS_NOTE_LEN>,
+        house: ImmutableSingleton<AddressNote, ADDRESS_NOTE_LEN>,
+        bets: Set<BetNote, BET_NOTE_LEN>,
     }
 
     impl Storage {
         fn init(context: Context) -> Self {
             Storage {
+                bet_amount: ImmutableSingleton::new(
```

Comment:
Maybe we can add all 4 pieces of data in just one note? It will make everything cleaner I think?